onnxruntime/core/providers/coreml/builders/impl/batch_norm_op_builder.cc

```diff
@@ -0,0 +1,139 @@
+#include "builder_utils.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class BatchNormalizationOpBuilder : public BaseOpBuilder {
+  // Add operator related
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const logging::Logger& logger) const override;
+
+  // BatchNormalization opset 6- has unsupported attributes
+  int GetMinSupportedOpSet(const Node& /* node */) const override { return 7; }
+};
+
+// Add operator related
+
+void BatchNormalizationOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  // skip everything except input0 for BatchNormalization
+  const auto& input_defs = node.InputDefs();
+  model_builder.AddInitializerToSkip(input_defs[1]->Name());  // scale
+  model_builder.AddInitializerToSkip(input_defs[2]->Name());  // B
+  model_builder.AddInitializerToSkip(input_defs[3]->Name());  // mean
+  model_builder.AddInitializerToSkip(input_defs[4]->Name());  // var
+}
+
+Status BatchNormalizationOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                                          const Node& node,
+                                                          const logging::Logger& /* logger */) const {
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(node);
+
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers(model_builder.GetInitializerTensors());
+  NodeAttrHelper helper(node);
+
+  const auto& scale_tensor = *initializers.at(input_defs[1]->Name());
+  const auto& bias_tensor = *initializers.at(input_defs[2]->Name());
+  const auto& mean_tensor = *initializers.at(input_defs[3]->Name());
+  const auto& var_tensor = *initializers.at(input_defs[4]->Name());
+  const auto eps = helper.Get("epsilon", 1e-5f);
+  const auto channels = scale_tensor.dims()[0];
+
+  auto* coreml_batch_norm = layer->mutable_batchnorm();
+  coreml_batch_norm->set_channels(channels);
+  coreml_batch_norm->set_epsilon(eps);
+  coreml_batch_norm->set_computemeanvar(false);
+  coreml_batch_norm->set_instancenormalization(false);
+
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_gamma(), scale_tensor);   // scale
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_beta(), bias_tensor);     // B
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_mean(), mean_tensor);     // mean
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_variance(), var_tensor);  // var
+
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(layer.release());
```

Comment:
>AddLayer [](start = 16, length = 8)

would it be better if ModelBuilder::AddLayer() takes a unique_ptr? then it conveys the transfer of ownership too